lib/srv/discovery/kube_watcher.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	concurrencyLimit = 5
+)
+
+func (s *Server) startKubeWatchers() error {
+	if len(s.kubeFetchers) == 0 {
+		return nil
+	}
+	var (
+		kubeResources types.ResourcesWithLabels
+		mu            sync.Mutex
+		t             = time.NewTicker(time.Minute)
+	)
+	watcher, err := services.NewReconciler(
+		services.ReconcilerConfig{
+			Matcher: func(_ types.ResourceWithLabels) bool { return true },
+			GetCurrentResources: func() types.ResourcesWithLabelsMap {
+				kcs, err := s.AccessPoint.GetKubernetesClusters(s.ctx)
+				if err != nil {
+					s.Log.WithError(err).Warnf("unable to get kubernetes clusters from cache")
+					return nil
+				}
+
+				// filter only discover clusters.
+				var kubeClusters types.KubeClusters
+				for _, kc := range kcs {
+					if kc.Origin() != types.OriginCloud {
+						continue
+					}
+					kubeClusters = append(kubeClusters, kc)
+				}
+
+				return kubeClusters.AsResources().ToMap()
+			},
+			GetNewResources: func() types.ResourcesWithLabelsMap {
+				mu.Lock()
+				defer mu.Unlock()
+				return kubeResources.ToMap()
+			},
+			Log:      s.Log,
+			OnCreate: s.onKubeCreate,
+			OnUpdate: s.onKubeUpdate,
+			OnDelete: s.onKubeDelete,
+		},
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+
+		for {
+
+			errGroup, errgCtx := errgroup.WithContext(s.ctx)
```

Comment:
Nit: first return arg has the `errgroup.Group` type so can we name it like bellow:  
```suggestion
			group, groupCtx := errgroup.WithContext(s.ctx)
```